File: lib/running.ts
/** Pure running math (client-safe) — roadmap branch 4. */

/**
 * Riegel race-time prediction: t2 = t1 * (d2/d1)^1.06.
 * Standard endurance exponent; good enough for goal suggestions.
 */
export function riegelSeconds(
  knownDistanceKm: number,
  knownSeconds: number,
  targetDistanceKm: number,
): number {
  if (knownDistanceKm <= 0 || knownSeconds <= 0 || targetDistanceKm <= 0) {
    return 0;
  }
  return Math.round(
    knownSeconds * Math.pow(targetDistanceKm / knownDistanceKm, 1.06),
  );
}

/** "4:05" → 245s, "3:59:30" → 14370s. Returns null when unparsable. */
export function parseTimeToSeconds(value: string): number | null {
  const parts = value
    .trim()
    .split(":")
    .map((part) => Number(part));
  if (parts.some((part) => !Number.isFinite(part) || part < 0)) return null;
  if (parts.length === 3) return parts[0] * 3600 + parts[1] * 60 + parts[2];
  if (parts.length === 2) return parts[0] * 60 + parts[1];
  return null;
}

/** 14370 → "3:59:30"; 245 → "4:05". */
export function formatSeconds(totalSeconds: number): string {
  const seconds = Math.max(0, Math.round(totalSeconds));
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const rest = seconds % 60;
  const mm = String(minutes).padStart(hours > 0 ? 2 : 1, "0");
  const ss = String(rest).padStart(2, "0");
  return hours > 0 ? `${hours}:${mm}:${ss}` : `${mm}:${ss}`;
}

export const RACE_DISTANCES_KM = {
  pb_5k: 5,
  pb_10k: 10,
  pb_half: 21.0975,
  pb_full: 42.195,
} as const;

/** Selectable race targets. Ultra/other use a user-typed distance. */
export const RACE_TARGETS = [
  { value: "5k", label: "5k", km: 5 },
  { value: "10k", label: "10k", km: 10 },
  { value: "half", label: "Half marathon", km: 21.0975 },
  { value: "full", label: "Marathon", km: 42.195 },
  { value: "ultra", label: "Ultra", km: null },
  { value: "other", label: "Other distance", km: null },
] as const;

export type RaceTarget = (typeof RACE_TARGETS)[number]["value"];

/** Resolve the race distance in km; ultra/other need customKm. */
export function raceDistanceKm(
  target: RaceTarget,
  customKm?: number | null,
): number | null {
  const preset = RACE_TARGETS.find((entry) => entry.value === target)?.km;
  if (preset) return preset;
  return typeof customKm === "number" && customKm > 0 ? customKm : null;
}

/**
 * Suggest a goal time for any race distance from the longest-distance PB
 * available (longer efforts predict better). Returns "h:mm:ss" or null.
 * Riegel degrades beyond ~50k, so ultra suggestions are rough at best.
 */
export function suggestGoalForDistance(
  pbs: {
    pb_5k?: string;
    pb_10k?: string;
    pb_half?: string;
    pb_full?: string;
  },
  targetKm: number,
): string | null {
  if (targetKm <= 0) return null;
  const order: (keyof typeof RACE_DISTANCES_KM)[] = [
    "pb_full",
    "pb_half",
    "pb_10k",
    "pb_5k",
  ];
  for (const key of order) {
    const raw = pbs[key];
    if (!raw) continue;
    const seconds = parseTimeToSeconds(raw);
    if (!seconds) continue;
    const knownKm = RACE_DISTANCES_KM[key];
    if (Math.abs(knownKm - targetKm) < 0.01) return formatSeconds(seconds);
    return formatSeconds(riegelSeconds(knownKm, seconds, targetKm));
  }
  return null;
}

/**
 * Suggest a marathon goal time from the longest-distance PB available
 * (longer efforts predict better). Returns formatted "h:mm:ss" or null.
 */
export function suggestMarathonGoal(pbs: {
  pb_5k?: string;
  pb_10k?: string;
  pb_half?: string;
  pb_full?: string;
}): string | null {
  return suggestGoalForDistance(pbs, RACE_DISTANCES_KM.pb_full);
}

File: apps/api/internal/domain/running/running.go
// Package running is the pure running math (FORMULAS.md §4–5): Riegel
// predictions, race distances, and time parsing/formatting.
package running

import (
	"fmt"
	"math"
	"strings"

	"github.com/hamid-karimi/coachin/apps/api/internal/domain/jsnum"
)

// RiegelExponent is the standard endurance fatigue exponent.
const RiegelExponent = 1.06

// RiegelSeconds predicts a time over targetKm from a known effort:
// t2 = t1 × (d2 / d1)^1.06, rounded to whole seconds. Non-positive inputs → 0.
func RiegelSeconds(knownKm, knownSeconds, targetKm float64) float64 {
	if knownKm <= 0 || knownSeconds <= 0 || targetKm <= 0 {
		return 0
	}
	return jsnum.Round(knownSeconds * math.Pow(targetKm/knownKm, RiegelExponent))
}

// ParseTimeToSeconds reads "m:ss" or "h:mm:ss" ("4:05" → 245,
// "3:59:30" → 14370). ok is false when it can't be read.
func ParseTimeToSeconds(value string) (seconds float64, ok bool) {
	raw := strings.Split(strings.TrimSpace(value), ":")
	parts := make([]float64, len(raw))
	for i, part := range raw {
		n, valid := jsnum.Number(part)
		if !valid || n < 0 {
			return 0, false
		}
		parts[i] = n
	}
	switch len(parts) {
	case 3:
		return parts[0]*3600 + parts[1]*60 + parts[2], true
	case 2:
		return parts[0]*60 + parts[1], true
	}
	return 0, false
}

// FormatSeconds renders whole seconds as "h:mm:ss", or "m:ss" under an hour.
func FormatSeconds(totalSeconds float64) string {
	seconds := int64(max(0, jsnum.Round(totalSeconds)))
	hours, minutes, rest := seconds/3600, (seconds%3600)/60, seconds%60
	if hours > 0 {
		return fmt.Sprintf("%d:%02d:%02d", hours, minutes, rest)
	}
	return fmt.Sprintf("%d:%02d", minutes, rest)
}

// Base-building ("just start running") program lengths, in weeks.
const (
	BaseWeeksDefault = 8
	BaseWeeksMin     = 4
	BaseWeeksMax     = 24
)

// BaseWeekOptions are the lengths offered in the wizard.
var BaseWeekOptions = []int{6, 8, 12}

// ClampBaseWeeks rounds and clamps a requested length to 4..24 weeks;
// missing input gets the default.
func ClampBaseWeeks(value *float64) int {
	if value == nil || math.IsNaN(*value) || math.IsInf(*value, 0) {
		return BaseWeeksDefault
	}
	return int(min(BaseWeeksMax, max(BaseWeeksMin, jsnum.Round(*value))))
}

// PBDistancesKm maps each personal-best key to its distance.
var PBDistancesKm = map[string]float64{
	"pb_5k":   5,
	"pb_10k":  10,
	"pb_half": 21.0975,
	"pb_full": 42.195,
}

// RaceTargetKm are the preset race targets; "ultra" and "other" take a
// user-typed distance instead.
var RaceTargetKm = map[string]float64{
	"5k":   5,
	"10k":  10,
	"half": 21.0975,
	"full": 42.195,
}

// RaceDistanceKm resolves a target's distance; ultra/other need a positive
// customKm. ok is false when there is no usable distance.
func RaceDistanceKm(target string, customKm *float64) (km float64, ok bool) {
	if preset, found := RaceTargetKm[target]; found {
		return preset, true
	}
	if customKm != nil && *customKm > 0 {
		return *customKm, true
	}
	return 0, false
}

// PBs are a runner's personal bests as typed ("25:00", "3:59:30").
type PBs struct {
	PB5k   string `json:"pb_5k"`
	PB10k  string `json:"pb_10k"`
	PBHalf string `json:"pb_half"`
	PBFull string `json:"pb_full"`
}

// SuggestGoalForDistance predicts a goal time for targetKm from the longest
// usable personal best (longer efforts predict better). ok is false when no
// PB can be read.
func SuggestGoalForDistance(pbs PBs, targetKm float64) (goal string, ok bool) {
	if targetKm <= 0 {
		return "", false
	}
	longestFirst := []struct{ key, time string }{
		{"pb_full", pbs.PBFull}, {"pb_half", pbs.PBHalf}, {"pb_10k", pbs.PB10k}, {"pb_5k", pbs.PB5k},
	}
	for _, pb := range longestFirst {
		if pb.time == "" {
			continue
		}
		seconds, valid := ParseTimeToSeconds(pb.time)
		if !valid || seconds == 0 {
			continue
		}
		knownKm := PBDistancesKm[pb.key]
		if math.Abs(knownKm-targetKm) < 0.01 {
			return FormatSeconds(seconds), true
		}
		return FormatSeconds(RiegelSeconds(knownKm, seconds, targetKm)), true
	}
	return "", false
}
